aip/auth/4110.md

```diff
@@ -0,0 +1,165 @@
+---
+aip:
+  id: 4110
+  scope: auth
+  state: reviewing
+  created: 2018-06-24
+  updated: 2019-06-24
+js:
+  - /assets/js/graphviz/viz.js
+  - /assets/js/graphviz/lite.render.js
+  - /assets/js/aip/aip-graphviz.js
+permalink: /auth/4110
+redirect_from:
+  - /4110
+---
+
+# Application Default Credentials
+
+Google auth libraries use a strategy called _Application Default Credentials
+(ADC)_ to detect and select credentials based on environment or context. With 
+ADC, developers should be able to run the code in different environments and the
+supporting systems fetch the appropriate credentials based on each environment
+in an effortless manner.
+
+Auth libraries following the standards in these AIPs are known as _"Google
+Unified Auth Clients"_, or _GUAC_ for short. The resulting libraries are
+colloquially called _GUACs_.
+
+**Note:** Because this AIP describes guidance and requirements in a
+language-neutral way, it uses generic terminology which may be imprecise or
+inappropriate in certain languages or environments.
+
+## Guidance
+
+### Credential Types
+
+This section outlines the supported credential types of the ADC.
+
+- **Gcloud Credential**: A credential provided by the [Gcloud tool][0] that
+identifies a human user that needs to authenticate to access Google APIs. The
+auth libraries **must** support this credential type.
+
+- **Service Account Key**: A credential that identifies a non-human user that
+needs to authenticate to aceess Google APIs. The auth libraries **must** support
+this credential type.
+
+- **OAuth Client ID**: A credential that identifies the client application which
+allows human users to sign-in through [3-legged OAuth flow][1], which grants the
+permissions to the application to access Google APIs on behave of the human
+user. The auth libraries **may** support this credential type.
+
+### Environment Variables
+
+The auth libraries **must** support the following environment variables to allow
+developers to provide authentication configuration for their application:
+
+- **GOOGLE_APPLICATION_CREDENTIALS**: The specified value will be used as the
+full path for ADC to locate the credentials file. The credentials file
+**should** be one of the following types:
+
+    - Service account key
+    - Gcloud credentials
+
+    The credentials **may** be the OAuth Client ID if it is supported by the
+    auth library. This value **must** be overridden by any user specified
+    credentials at the programming level (e.g. through client-level options)
+
+- **GOOGLE_API_KEY**: [API keys][2] are a simple encrypted string that
+identifies an application, which are useful for accessing public data
+anonymously. The specified value of this environment variable **should** be used
+as the API key for the API requests. This value **must** be overridden by any
+user specified credentials at the programming level (e.g. through client
+options). And it **must** be ignored if GOOGLE_APPLICATION_CREDENTIALS is
```

Comment:
Similar to above. Saying that "something must be overridden" suggests we literally must always provide client options to override this value. The intent we want to communicate should be that "client options has priority over environment variables."